src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -3362,12 +3369,44 @@ class CyclesRenderer final : public IECoreScenePreview::Renderer
 			m_shaderCache->update( m_scene, m_shadersCreated );
 		}
 
-		void updateOptions()
+		void updateSessionParams()
 		{
 #ifdef WITH_CYCLES_TEXTURE_CACHE
 			m_sceneParams.texture = m_textureCacheParams;
 #endif
+			// Check if an OSL shader exists & set the shadingsystem
+			if( m_sessionParams.shadingsystem == ccl::SHADINGSYSTEM_SVM && m_shaderCache->hasOSLShader() )
+			{
+				IECore::msg( IECore::Msg::Warning, "CyclesRenderer", "OSL Shader detected, forcing OSL shading-system (CPU-only)" );
+				m_sessionParams.shadingsystem = ccl::SHADINGSYSTEM_OSL;
+				m_sceneParams.shadingsystem = ccl::SHADINGSYSTEM_OSL;
+			}
+
+			// If anything changes in scene or session, we reset.
+			if( m_scene->params.modified( m_sceneParams ) ||
+				m_session->params.modified( m_sessionParams )
+			)
+			{
+				copySocketValues( *(m_scene->integrator), m_integrator );
+				copySocketValues( *(m_scene->background), m_background );
+				copySocketValues( *(m_scene->film), m_film );
 
+				init();
```

Comment:
It only does it before rendering starts and not during rendering. There's still roadblocks to consider before moving the `init()` to `render()`.

Reply:
So the tricky to reuse the memory address is redundant now, and can be removed?

Reply:
If there's a way to avoid these tags I think so:
https://github.com/GafferHQ/gaffer/blob/main/src/GafferCycles/IECoreCyclesPreview/Renderer.cpp#L2009

A forward declare of `CyclesRender` and giving it a session or scene getter didn't work, not too sure on what other solutions to have here - is there another way to find out if an object got a transform or attribute change from the `Renderer`?

Reply:
Ok I see that nodes have a `is_modified()` now which can defer to the `CyclesRenderer`, that's good.